File: src/main/java/net/cakemc/database/collection/PieceCollection.java
package net.cakemc.database.collection;

import net.cakemc.database.AbstractDatabase;
import net.cakemc.database.api.DatabaseRecord;
import net.cakemc.database.api.Piece;
import net.cakemc.database.cursor.Cursor;
import net.cakemc.database.cursor.PieceCursorSupplier;
import net.cakemc.database.callbacks.DatabaseListener;
import net.cakemc.database.callbacks.PieceAsyncCallBack;
import net.cakemc.database.callbacks.PieceAsyncMultiCallBack;
import net.cakemc.database.filter.PieceFilter;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.List;
import java.util.concurrent.CompletableFuture;
import java.util.concurrent.ExecutionException;
import java.util.concurrent.ThreadLocalRandom;

/**
 * The type Piece collection.
 */
public class PieceCollection extends AbstractCollection<DatabaseRecord> {


    /**
     * Instantiates a new Piece collection.
     *
     * @param elements the elements
     * @param id       the id
     * @param name     the name
     */
    public PieceCollection(List<DatabaseRecord> elements, long id, String name) {
        super(elements, id, name);
    }


    @Override
    public void multiPieceAsync(PieceFilter filter, PieceCursorSupplier cursorSupplier, PieceAsyncMultiCallBack callBack) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                callBack.acceptFound(cursorSupplier.create(this.elements.stream()
                        .filter(databaseRecord -> filter.matches((Piece) databaseRecord))
                        .map(databaseRecord -> (Piece) databaseRecord)
                        .toList()));

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            callBack.acceptException(exception);
        }
    }

    @Override
    public void singlePieceAsync(PieceFilter filter, PieceAsyncCallBack documentCallBack) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                Piece element = (Piece) elements.stream()
                        .filter(databaseRecord -> filter.matches((Piece) databaseRecord))
                        .findFirst()
                        .orElse(null);

                if (element == null) {
                    documentCallBack.acceptNotFound();
                    return;
                }

                documentCallBack.acceptFound(element);

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            documentCallBack.acceptException(exception);
        }
    }

    @Override
    public Cursor<Piece> multiPiece(PieceFilter filter, PieceCursorSupplier cursorSupplier) {
        return cursorSupplier.create(elements.stream()
                .filter(databaseRecord -> filter.matches((Piece) databaseRecord))
                .map(databaseRecord -> (Piece) databaseRecord).toList());
    }

    @Override
    public Piece singlePiece(PieceFilter filter) {
        return (Piece) elements.stream().filter(databaseRecord -> filter.matches((Piece) databaseRecord)).findFirst().orElse(null);
    }

    @Override
    public void replaceOnePiece(PieceFilter filter, Piece element) {
        elements.removeIf(databaseRecord -> filter.matches((Piece) databaseRecord));
        elements.add(element);
    }

    @Override
    public void replaceOnePieceAsync(PieceFilter filter, Piece element, DatabaseListener listener) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                if (elements.removeIf(databaseRecord -> filter.matches((Piece) databaseRecord))) {
                    elements.add(element);
                    listener.success();
                } else
                    listener.accept(DatabaseListener.State.FAILED, null);

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            listener.exception(exception);
        }
    }

    @Override
    public void updateOnePiece(PieceFilter filter, Piece element) {
        // todo update fields in filter
    }

    @Override
    public void updateOnePieceAsync(PieceFilter filter, Piece element, DatabaseListener listener) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                // todo update fields in filter

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            listener.exception(exception);
        }
    }

    @Override
    public void deleteOnePiece(Piece element) {
        this.elements.remove(element);
    }

    @Override
    public void deleteManyPiece(Piece... element) {
        for (Piece piece : element) {
            this.elements.remove(piece);
        }

    }

    @Override
    public void deleteOnePieceAsync(Piece element, DatabaseListener listener) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                if (elements.remove(element))
                    listener.success();
                else
                    listener.accept(DatabaseListener.State.FAILED, null);

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            listener.exception(exception);
        }
    }

    @Override
    public void deleteManyPieceAsync(DatabaseListener listener, Piece... element) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                for (DatabaseRecord current : element) {
                    elements.remove(current);
                }

                listener.success();

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            listener.exception(exception);
        }
    }

    @Override
    public void insertOnePiece(Piece element) {
        this.elements.add(element);
    }

    @Override
    public void insertOnePieceAsync(Piece element, DatabaseListener listener) {
        try {
            CompletableFuture<Void> completableFuture = CompletableFuture.runAsync(() -> {

                this.elements.add(element);
                listener.success();

            }, AbstractDatabase.VISUAL_EXECUTOR);
            completableFuture.get();
        } catch (InterruptedException | ExecutionException exception) {
            listener.exception(exception);
        }
    }

    private <T> Piece pieceCast(T element) {
        return (Piece) element;
    }

    @Override
    public DatabaseRecord defineOne() {
        throw new UnsupportedOperationException("no implemented!");
    }

    @Override
    public Piece defineOnePeace() {
        int index = this.elements.size() + 1;
        long id = nextFreeId();

        return new Piece(index, id, new HashMap<>());
    }

    private long nextFreeId() {
        long current = ThreadLocalRandom.current().nextLong();

        if (this.elements.stream()
                .anyMatch(databaseRecord -> databaseRecord.getId() == current))

            return nextFreeId();
        return current;
    }

    @Override
    public List<Piece> collect() {
        List<Piece> pieces = new ArrayList<>();
        for (DatabaseRecord element : this.elements) {
            if (element instanceof Piece piece)
                pieces.add(piece);
        }
        return pieces;
    }
}

File: src/main/kotlin/net/cakemc/database/collection/PieceCollection.kt
package net.cakemc.database.collection

import net.cakemc.database.AbstractDatabase
import net.cakemc.database.api.DatabaseRecord
import net.cakemc.database.api.Piece
import net.cakemc.database.callbacks.DatabaseListener
import net.cakemc.database.callbacks.PieceAsyncCallBack
import net.cakemc.database.callbacks.PieceAsyncMultiCallBack
import net.cakemc.database.cursor.Cursor
import net.cakemc.database.cursor.PieceCursorSupplier
import net.cakemc.database.filter.PieceFilter
import java.util.concurrent.CompletableFuture
import java.util.concurrent.ExecutionException
import java.util.concurrent.ThreadLocalRandom

/**
 * Piece collection implementation.
 */
class PieceCollection(
    elements: MutableList<DatabaseRecord>,
    id: Long,
    name: String
) : AbstractCollection<DatabaseRecord>(elements, id, name) {

    override fun multiPieceAsync(filter: PieceFilter, cursorSupplier: PieceCursorSupplier, callBack: PieceAsyncMultiCallBack) {
        try {
            CompletableFuture.runAsync({
                val matchedElements = elements
                    .filter { filter.matches(it as Piece) }
                    .map { it as Piece }

                callBack.acceptFound(cursorSupplier.create(matchedElements))
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            callBack.acceptException(exception)
        } catch (exception: ExecutionException) {
            callBack.acceptException(exception)
        }
    }

    override fun singlePieceAsync(supplier: PieceFilter, documentCallBack: PieceAsyncCallBack) {
        try {
            CompletableFuture.runAsync({
                val element = elements
                    .filterIsInstance<Piece>()
                    .find { supplier.matches(it) }

                if (element == null) {
                    documentCallBack.acceptNotFound()
                } else {
                    documentCallBack.acceptFound(element)
                }
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            documentCallBack.acceptException(exception)
        } catch (exception: ExecutionException) {
            documentCallBack.acceptException(exception)
        }
    }

    override fun multiPiece(filter: PieceFilter, cursorSupplier: PieceCursorSupplier): Cursor<Piece> {
        val matchedElements = elements
            .filterIsInstance<Piece>()
            .filter { filter.matches(it) }
        return cursorSupplier.create(matchedElements)
    }

    override fun singlePiece(supplier: PieceFilter): Piece? {
        return elements.filterIsInstance<Piece>().find { supplier.matches(it) }
    }

    override fun replaceOnePiece(filter: PieceFilter, element: Piece) {
        elements.removeIf { filter.matches(it as Piece) }
        elements.add(element)
    }

    override fun replaceOnePieceAsync(filter: PieceFilter, element: Piece, listener: DatabaseListener) {
        try {
            CompletableFuture.runAsync({
                if (elements.removeIf { filter.matches(it as Piece) }) {
                    elements.add(element)
                    listener.success()
                } else {
                    listener.accept(DatabaseListener.State.FAILED, null)
                }
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            listener.exception(exception)
        } catch (exception: ExecutionException) {
            listener.exception(exception)
        }
    }

    override fun updateOnePiece(filter: PieceFilter, other: Piece) {
        val current = this.elements.stream().filter { filter.matches(it as Piece) }.findFirst()

        if (current.isEmpty) {
            this.insertOnePiece(other)
            return
        }

        val piece: Piece = current.get() as Piece;

        for (containing in other.elements) {
            val key = containing.key;

            if (piece.contains(key)) {
                piece.remove(key)
                piece.set(containing.key, containing.value)
                continue
            }

            piece.set(containing.key, containing.value)
        }

        elements.removeIf { filter.matches(it as Piece) }
        elements.add(piece)
     }

    override fun updateOnePieceAsync(filter: PieceFilter, element: Piece, listener: DatabaseListener) {
        try {
            CompletableFuture.runAsync({
                updateOnePiece(filter, element)
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            listener.exception(exception)
        } catch (exception: ExecutionException) {
            listener.exception(exception)
        }
    }

    override fun deleteOnePiece(element: Piece) {
        elements.remove(element)
    }

    override fun deleteManyPiece(vararg element: Piece) {
        element.forEach { elements.remove(it) }
    }

    override fun deleteOnePieceAsync(element: Piece, listener: DatabaseListener) {
        try {
            CompletableFuture.runAsync({
                if (elements.remove(element)) {
                    listener.success()
                } else {
                    listener.accept(DatabaseListener.State.FAILED, null)
                }
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            listener.exception(exception)
        } catch (exception: ExecutionException) {
            listener.exception(exception)
        }
    }

    override fun deleteManyPieceAsync(listener: DatabaseListener, vararg element: Piece) {
        try {
            CompletableFuture.runAsync({
                element.forEach { elements.remove(it) }
                listener.success()
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            listener.exception(exception)
        } catch (exception: ExecutionException) {
            listener.exception(exception)
        }
    }

    override fun insertOnePiece(element: Piece) {
        elements.add(element)
    }

    override fun insertOnePieceAsync(element: Piece, listener: DatabaseListener) {
        try {
            CompletableFuture.runAsync({
                elements.add(element)
                listener.success()
            }, AbstractDatabase.EXECUTOR).get()
        } catch (exception: InterruptedException) {
            listener.exception(exception)
        } catch (exception: ExecutionException) {
            listener.exception(exception)
        }
    }

    override fun defineOne(): DatabaseRecord {
        throw UnsupportedOperationException("Not implemented!")
    }

    override fun defineOnePeace(): Piece {
        val index = elements.size + 1
        val id = nextFreeId()

        val data = mutableMapOf<String, Any>(
            "_created_at" to System.currentTimeMillis(),
            "_auto_delete" to -1
        )

        return Piece(index, id, data)
    }

    private fun nextFreeId(): Long {
        val current = ThreadLocalRandom.current().nextLong()
        return if (elements.any { it.id == current }) nextFreeId() else current
    }

    override fun collect(): List<Piece> {
        return elements.filterIsInstance<Piece>()
    }
}
